File: gren/v1/kit/graphs/reducers/triage-result.js
// Deterministic per-ticket result: needs_human is computed by code from the route, the policy check and the verifier.
// input: { ticket, label, trail, route, reply_survivors, reply_killed, policy }
export default async function triageResult(input) {
  const survivors = Array.isArray(input.reply_survivors) ? input.reply_survivors : [];
  const killed = Array.isArray(input.reply_killed) ? input.reply_killed : [];
  const policy = input.policy ?? {};
  const escalated = input.route === "escalate";
  const reply = survivors[0] ?? null;
  const reasons = [];
  if (escalated) reasons.push(`routed to escalate (label ${input.label?.label ?? input.label ?? "?"})`);
  if (policy.needs_human) reasons.push(...(policy.violations?.length ? policy.violations : ["policy check requires a human"]));
  if (!escalated && !reply) reasons.push(killed.length ? `reply rejected by verifier: ${(killed[0]?.reasons ?? []).join("; ")}` : "no reply drafted");
  return {
    ticket_id: input.ticket?.id ?? null,
    customer: input.ticket?.customer ?? null,
    tier: input.ticket?.tier ?? null,
    label: input.label?.label ?? input.label ?? null,
    confidence: input.label?.confidence ?? null,
    route: input.route ?? null,
    reply,
    refund_usd: reply?.refund_usd ?? 0,
    needs_human: reasons.length > 0,
    reasons,
    trail: input.trail ?? {},
    policy,
  };
}

File: kit/graphs/reducers/triage-result.py
# Deterministic per-ticket result: needs_human is computed by code from the route, the policy check and the verifier.
# input: { ticket, label, trail, route, reply_survivors, reply_killed, policy }


def reduce(input, args, ctx):
    survivors = input.get("reply_survivors") if isinstance(input.get("reply_survivors"), list) else []
    killed = input.get("reply_killed") if isinstance(input.get("reply_killed"), list) else []
    policy = input.get("policy") or {}
    label = input.get("label")
    label_value = label.get("label") if isinstance(label, dict) else label
    confidence = label.get("confidence") if isinstance(label, dict) else None
    escalated = input.get("route") == "escalate"
    reply = survivors[0] if survivors else None
    reasons: list[str] = []
    if escalated:
        reasons.append(f"routed to escalate (label {label_value or '?'})")
    if policy.get("needs_human"):
        reasons.extend(policy.get("violations") or ["policy check requires a human"])
    if not escalated and reply is None:
        reasons.append(f"reply rejected by verifier: {'; '.join((killed[0].get('reasons') or []) if isinstance(killed[0], dict) else [])}" if killed else "no reply drafted")
    ticket = input.get("ticket") or {}
    return {
        "ticket_id": ticket.get("id"), "customer": ticket.get("customer"), "tier": ticket.get("tier"), "label": label_value, "confidence": confidence,
        "route": input.get("route"), "reply": reply, "refund_usd": (reply.get("refund_usd") if isinstance(reply, dict) else None) or 0,
        "needs_human": bool(reasons), "reasons": reasons, "trail": input.get("trail") or {}, "policy": policy,
    }
